### PR TITLE
Update Gateway API to v0.7.0

### LINKED ISSUE #25474: Upgrade Gateway API to 0.7.0
Gateway API v0.7.0 was released (https://github.com/kubernetes-sigs/gateway-api/releases/tag/v0.7.0)
While this was not a big change regarding the spec or features there are some conformance changes which we should work on. 

Items to cover as part of this issue
- [x] Update all dependencies to v0.7.0
- [x] Make required fixes to pass conformance
- [ ] Check disabled conformance tests to see if they can be re-enabled

### PR DESCRIPTION
**Pull request by meyskens**
This PR updates our Gateway API to v0.7.0. This includes the new vendor changes as well as all fixes we needed to pass the additional conformance tests in individual commits.

Fixes: https://github.com/cilium/cilium/issues/25474

```release-note
Support Gateway API v0.7.0
```

### CODE REVIEW

File: operator/pkg/gateway-api/httproute.go [hunk 1]
```diff
@@ -92,6 +93,8 @@ func (r *httpRouteReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&gatewayv1beta1.HTTPRoute{}).
 		// Watch for changes to Backend services
 		Watches(&source.Kind{Type: &corev1.Service{}}, r.enqueueRequestForBackendService()).
+		// Watch for changes to Reference Grants
+		Watches(&source.Kind{Type: &gatewayv1beta1.ReferenceGrant{}}, r.enqueueRequestForRequestGrant()).
```
Comment (sayboras) [inline operator/pkg/gateway-api/httproute.go:97]: :100:

**Review by sayboras — APPROVED** [review]
Thanks for working on this, also cleaning up my bad code/mistake before :+1:

File: operator/pkg/gateway-api/gateway_status.go [hunk 2]
```diff
@@ -45,6 +51,29 @@ func gatewayStatusAcceptedCondition(gw *gatewayv1beta1.Gateway, accepted bool, m
 	}
 }
 
+func gatewayStatusProgrammedCondition(gw *gatewayv1beta1.Gateway, scheduled bool, msg string) metav1.Condition {
+	switch scheduled {
```
Comment (viktor-kurchenko) [inline operator/pkg/gateway-api/gateway_status.go:49]: Just wondering, why `switch` not `if`?

File: operator/pkg/gateway-api/helpers/referencegrants.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+package helpers
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	schema "k8s.io/apimachinery/pkg/runtime/schema"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+// isReferenceAllowed returns true if the reference is allowed by the reference grant.
```
Comment (viktor-kurchenko) [inline operator/pkg/gateway-api/helpers/referencegrants.go]: `isReferenceAllowed` -> `IsBackendReferenceAllowed`

File: operator/pkg/gateway-api/gateway_status.go (hunk 2 above)
Comment (meyskens) [reply operator/pkg/gateway-api/gateway_status.go:49]: This is an adaptation from @sayboras's original code, not sure what his reasoning was :)

File: operator/pkg/gateway-api/helpers/referencegrants.go (hunk 3 above)
Comment (meyskens) [reply operator/pkg/gateway-api/helpers/referencegrants.go]: Will change this in the next PR where i also add checks for certificates

File: operator/pkg/gateway-api/gateway_status.go (hunk 2 above)
Comment (sayboras) [reply operator/pkg/gateway-api/gateway_status.go:49]: ah i might have started with some enum, and then refactored to bool, so switch might be just lazy refactor.

Comment (meyskens) [reply operator/pkg/gateway-api/gateway_status.go:49]: I have had an idea last night to re-factor this into 1 generic blob to re-use. But that is probably out of scope for this PR.

**Review by viktor-kurchenko — APPROVED** [review]

File: operator/pkg/gateway-api/gateway_reconcile.go [hunk 4]
```diff
@@ -143,14 +144,16 @@ func (r *gatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return fail(err)
 	}
 
+	setGatewayReady(gw, true, "Gateway successfully configured")
```
Comment (youngnick) [inline operator/pkg/gateway-api/gateway_reconcile.go]: Hmm, Ready is now deprecated, actually, so we shouldn't set it.

File: operator/pkg/gateway-api/gateway_reconcile.go [hunk 5]
```diff
@@ -319,8 +349,10 @@ func (r *gatewayReconciler) setListenerStatus(ctx context.Context, gw *gatewayv1
 		var conds []metav1.Condition
 		if invalidRouteKinds {
 			conds = append(conds, gatewayListenerInvalidRouteKinds(gw, "Invalid Route Kinds"))
+			isValid = false
 		} else {
 			conds = append(conds, gatewayListenerProgrammedCondition(gw, true, "Listener Ready"))
```
Comment (youngnick) [inline operator/pkg/gateway-api/gateway_reconcile.go]: Ready has been deprecated as part of this release, so we should remove it.

**Comment by sayboras** [comment]
Add a preview-only label as we might need to wait for https://github.com/cilium/cilium/pull/25602 to land in first.

**Review by youngnick — APPROVED** [review]
LGTM with one small thing about removing `Ready` if we can. It's only deprecated, so not a blocker, but if we can remove now, that would be better.

File: operator/pkg/gateway-api/gateway_reconcile.go (hunk 4 above)
Comment (meyskens) [reply operator/pkg/gateway-api/gateway_reconcile.go]: Removing it :)

**Comment by sayboras** [comment]
/test-1.26-net-next

**Comment by sayboras** [comment]
Rebased with master to get the fix for ci-l4lb

**Comment by sayboras** [comment]
/test-1.16-4.19

**Pull request merged**